File: preparer/src/bin/brotler.rs
use std::{env, process};
use std::fmt::{Display, Formatter};
use std::fs::File;
use std::io::copy;
use std::path::Path;
use brotlic::{BrotliEncoderOptions, CompressorWriter, Quality, SetParameterError, WindowSize};
use log::{debug, error};
use tempfile::{NamedTempFile, TempPath};
use walkdir::WalkDir;

#[derive(Debug)]
enum MyError {
    IO(std::io::Error),
    WalkDir(walkdir::Error),
    SetParameter(SetParameterError),
}

impl Display for MyError {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::IO(err) => write!(f, "I/O: {err}"),
            Self::WalkDir(err) => write!(f, "Walking dir: {err}"),
            Self::SetParameter(err) => write!(f, "Setting parameter: {err}"),
        }
    }
}

impl From<walkdir::Error> for MyError {
    fn from(value: walkdir::Error) -> Self {
        Self::WalkDir(value)
    }
}

impl From<std::io::Error> for MyError {
    fn from(value: std::io::Error) -> Self {
        Self::IO(value)
    }
}

impl From<SetParameterError> for MyError {
    fn from(value: SetParameterError) -> Self {
        Self::SetParameter(value)
    }
}

fn main() {
    env_logger::builder().init();
    if let Err(err) = almost_main() {
        error!("{err}");
        process::exit(1);
    }
}

fn almost_main() -> Result<(), MyError> {
    if env::args().len() != 2 {
        error!("Usage: brotler <DIR>");
        process::exit(1);
    }
    for entry in WalkDir::new(Path::new(&env::args().nth(1).unwrap()))
        .sort_by_file_name() // keep the order deterministic, because we overwrite files
        .into_iter()
        .filter_entry(|entry| !entry.path_is_symlink())
    {
        let entry = entry?;
        if !entry.file_type().is_dir() {
            debug!("Compressing file {}", entry.path().to_str().unwrap());
            compress_file(&entry.path())?;
        }
    }

    Ok(())
}

fn compress_file(path: &Path) -> Result<(), MyError> {
    let mut input = File::open(path.clone())?; // uncompressed text file
    let output_file = NamedTempFile::new()?;
    let output_path: TempPath = output_file.into_temp_path();
    let output_path: &Path = output_path.as_ref();
    let output = File::create(output_path)?; // compressed text output file

    let encoder = BrotliEncoderOptions::new()
        .quality(Quality::best())
        .window_size(WindowSize::new(24).unwrap())
        .build()?;

    let mut output_compressed = CompressorWriter::with_encoder(encoder, output);

    copy(&mut input, &mut output_compressed)?;
    drop(output_compressed);
    // rename(output_path, path)?; // /tmp is on another filesystem
    std::fs::copy(output_path, path)?;
    Ok(())
}

File: preparer/src/brotler.go
package main

import (
	"io"
	"io/fs"
	"io/ioutil"
	"log"
	"os"
	"path/filepath"

	"github.com/andybalholm/brotli"
)

func main() {
	logger := log.Default()
	if err := almost_main(logger); err != nil {
		logger.Fatal(err)
	}
}

func almost_main(logger *log.Logger) error {
	if len(os.Args) != 2 {
		logger.Fatal("Usage: brotler <DIR>")
	}
	filepath.WalkDir(os.Args[1], func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return err
		}
		info, err2 := d.Info()
		if err2 != nil {
			return err2
		}
		if info.Mode()&os.ModeSymlink == 0 && !info.IsDir() {
			logger.Print("Compressing file ", path)
			if err := compressFile(path); err != nil {
				return err
			}
		}
		return nil
	})
	return nil
}

func compressFile(path string) error {
	input, err0 := os.Open(path)
	if err0 != nil {
		return err0
	}
	output, err := ioutil.TempFile("/tmp", "brotler")
	if err != nil {
		return err
	}

	encoder := brotli.NewWriterLevel(output, brotli.BestCompression)
	io.Copy(encoder, input)
	encoder.Close()
	output_rev, err2 := os.Open(output.Name())
	if err2 != nil {
		return err2
	}
	input_rev, err3 := os.Create(path)
	if err3 != nil {
		return err3
	}
	io.Copy(input_rev, output_rev)
	return nil
}
